README.md
# NAME

App::Nopaste - easy access to any pastebin

# VERSION

version 0.96

# SYNOPSIS

    use App::Nopaste 'nopaste';

    my $url = nopaste(q{
        perl -wle 'print "Prime" if (1 x shift) !~ /^1?$|^(11+?)\1+$/' [number]
    });

    # or on the command line:
    nopaste test.pl
    => http://pastebin.com/fcba51f

# DESCRIPTION

Pastebins (also known as nopaste sites) let you post text, usually code, for
public viewing. They're used a lot in IRC channels to show code that would
normally be too long to give directly in the channel (hence the name nopaste).

Each pastebin is slightly different. When one pastebin goes down (I'm looking
at you, [http://paste.husk.org](http://paste.husk.org)), then you have to find a new one. And if you
usually use a script to publish text, then it's too much hassle.

This module aims to smooth out the differences between pastebins, and provides
redundancy: if one site doesn't work, it just tries a different one.

It's also modular: you only need to put on CPAN a
[App::Nopaste::Service::Foo](https://metacpan.org/pod/App::Nopaste::Service::Foo) module and anyone can begin using it.

# INTERFACE

## CLI

See the documentation in [App::Nopaste::Command](https://metacpan.org/pod/App::Nopaste::Command).

## `nopaste`

    use App::Nopaste 'nopaste';

    my $url = nopaste(
        text => "Full text to paste (the only mandatory argument)",
        desc => "A short description of the paste",
        nick => "Your nickname",
        lang => "perl",
        chan => "#moose",
        private => 1, # default: 0

        # this is the default, but maybe you want to do something different
        error_handler => sub {
            my ($error, $service) = @_;
            warn "$service: $error";
        },

        warn_handler => sub {
            my ($warning, $service) = @_;
            warn "$service: $warning";
        },

        # you may specify the services to use - but you don't have to
        services => ["Shadowcat", "Gist"],
    );

    print $url if $url;

The `nopaste` function will return the URL of the paste on
success, or `undef` on failure.

For each failure, the `error_handler` argument is invoked with the error
message and the service that issued it.

For each warning, the `warn_handler` argument is invoked with the warning
message and the service that issued it.

# SEE ALSO

[WebService::NoPaste](https://metacpan.org/pod/WebService::NoPaste), [WWW::Pastebin::PastebinCom::Create](https://metacpan.org/pod/WWW::Pastebin::PastebinCom::Create), [Devel::REPL::Plugin::Nopaste](https://metacpan.org/pod/Devel::REPL::Plugin::Nopaste)

[http://perladvent.org/2011/2011-12-14.html](http://perladvent.org/2011/2011-12-14.html)

# AUTHOR

Shawn M Moore, `sartak@gmail.com`

# COPYRIGHT AND LICENSE

Copyright 2008- Shawn M Moore.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.96      2014-01-29 03:24:58Z
        - added missing prereqs

0.95      2014-01-26 17:41:29Z
        - bumped the minimum version needed for the (optional) prerequisite
          WWW::Pastebin::PastebinCom::Create, to avoid errors in previous
          versions (thanks, Zoffix!)

0.94      2014-01-18 22:13:29Z
        - updated and fixed debian and ubuntu language mappings (haarg, PR#1)

0.93      2014-01-14 00:41:21Z
        - packaging converted to Dist::Zilla; handling of optional features
          fixed (see RT#91053)

0.92    2013-04-26
        Fix the language mappins for Ubuntu's pastebin (Sergey Romanov)
        Don't need to load Moose in the tests (Sergey Romanov)

0.91    2013-01-27
        Decode incoming text as utf8 (Jesse Luehrs)
        Support for GITHUB_USER and GITHUB_PASS (Rafael Kitover)
        Name the gist file `nopaste` not `nopaste.txt` for possible syntax highlighting (Rafael Kitover)
        Use decode_json not a regular expression to decode Gist response body (Shawn M Moore)
        Give a hint to perldoc App::Nopaste::Service::Gist on 401 (Shawn M Moore)

0.90    2012-10-30
        Port from Moose and MooseX::Getopt to Getopt::Long::Descriptive (frew)
            App::Nopaste is not using any interesting Moose features,
            so it doesn't make sense to pay the startup cost (even
            though it's probably dominated by network latency)

            Please report any regressions!

0.35    2012-07-17
        Don't barf when gist-ing STDIN. (Ricardo SIGNES)

0.34    2012-07-14
        Support for GitHub's v3 API [Вячеслав Тихановский (vti)]
            Thanks also to Ricardo SIGNES for the call to arms:
                http://rjbs.manxome.org/rubric/entry/1964

0.33    2011-12-15
        Support for --description in Gist (zengargoyle@gmail.com) [rt.cpan.org #73272]

0.32    2011-10-21
        Better error reporting (David Golden)
        Fix some attribution errors (Salvatore Bonaccorso) [rt.cpan.org #70434]

0.31    2011-08-30
        Better filenames for ssh service (Thomas Sibley)

0.30    2011-08-26
        Fix App::Nopaste::ssh's new filename description feature (Kevin Falcone)

0.29    2011-08-22
        Add --private to several services (François Gannaz)
        Use Class::Load instead of require "Module/Name.pm" (Shawn M Moore)
        Document --list and --private options (François Gannaz)
        Less hardcoding of argument names to better support subclassing (Jason Mills)
        New Ubuntu service (http://paste.ubuntu.com/) (gregoa@debian.org)
        SSH: Add a small suggested Apache conf that's useful for paste directories (Thomas Sibley)
        SSH: optionally use the supplied description in the paste filename (Thomas Sibley)

0.28    2011-03-22
        If LWP is producing errors, *report them* (Ricardo SIGNES)

0.27    2011-03-10
        Correct path to Pastie (Sebastian Paaske Tørholm )
        Throw an error if you specify -p and files (Shawn M Moore)
        Remove Mathbin; doy moved it to a separate dist (Shawn M Moore)

0.26    2011-02-23
        Add --open (-o) for opening the nopaste in your browser (Thomas Sibley)

0.25    2011-01-03
        Add support for $GITHUB_USER/$GITHUB_TOKEN to Gist service (Maximilian Gass)

0.24    2010-12-21
        Gist requires https (Ricardo SIGNES)

0.23    2010-11-26
        Preserve the source file's extension for ssh (Thomas Sibley)
        Use Config::GitLike instead of Config::INI::Reader (Thomas Sibley)
        Doc fixes (Thomas Sibley, Justin Hunter)
        Code style fixes (Shawn M Moore)
        Remove Rafb (Justin Hunter)
        Remove Husk (Shawn M Moore)

0.22    2010-06-10 12:58:00
        Add NOPASTE_SSH_MODE for chmod'ing after uploading (Thomas Sibley)

0.21    2010-05-01 09:32:10
        Add a --list_services/-L option (David Bremner) [rt.cpan.org #55562]
        Add --private support, used in Gist (John Goulah)
        Doc fix (Salvatore Bonaccorso) [rt.cpan.org #53249]
        Doc fix (Fabien Wernli) [rt.cpan.org #54928]
        Mention 'nopaste' script (Damyan Ivanov) [rt.cpan.org #51065]

0.20    2010-04-16 23:05:10
        App::Nopaste::Service::Pastie would generate warnings via
            Mechanize due to changes in http://pastie.org HTML (avar)
        Add App::Nopaste::Service::Codepeek (arcanez)

0.19    2010-02-21 15:39:27
        Remove rafb.net support since it's been shut down
            - See also http://rafb.net/paste/

0.18    2009-12-30 16:04:45
        Add ssh support (obra, jibsheet)

0.17    2009-11-02 13:20:24
        fix Gist support harder by using the API rather than scraping (rjbs)
        Improve --lang support for Debian (gregoa@debian.org) [rt.cpan.org #47911]

0.16    2009-10-16 14:32:09
        Fix Gist support (ckuskie@sterling.net) [rt.cpan.org #50500]
        Add --lang support to Debian (gregoa@debian.org) [rt.cpan.org #47911]

0.15    2009-07-15 05:01:36
        MANIFEST is really hard! :(

0.14    2009-07-15 04:48:15
        No code changes

0.13    2009-07-15 04:47:21
        Add support for paste.debian.net (Ryan Niebur)

0.12    2009-07-13 19:35:10
        Don't explicitly (optionally) depend on Git, but use it if
            we have it (Sartak)

0.11    2009-06-18 17:20:51
        Service::Gist: use file_ext instead of file_name (nothingmuch)

        Bump Moose dependencies to avoid deprecation warnings

        Doc typo fix (rafl)

0.10    2009-01-28 18:48:55
        I was a little too aggressive in my s/user/login/ (Sartak)

0.09    2009-01-28 18:34:31
        The parameter for gist is "user" not "login" (Ricardo SIGNES)

0.08    2009-01-22 21:44:51
        Optional authentication for Gist, by Ricardo SIGNES.

0.07    2009-01-16 16:04:45
        Gist (github) support by Ricardo SIGNES.
        Pastie support.

0.06    2008-12-03 21:50:36
        Silence the "exiting eval via next" warning

0.05    2008-08-29 15:03:10

lib/App/Nopaste/Service/Pastie.pm
use strict;
use warnings;
package App::Nopaste::Service::Pastie;

use base 'App::Nopaste::Service';

my %languages = (
    "bash" => "13",
    "c#" => "20",
    "c/c++" => "7",
    "css" => "8",
    "diff" => "5",
    "go" => "21",
    "html (erb / rails)" => "12",
    "html / xml" => "11",
    "java" => "9",
    "javascript" => "10",
    "objective-c/c++" => "1",
    "perl" => "18",
    "php" => "15",
    "plain text" => "6",
    "python" => "16",
    "ruby" => "3",
    "ruby on rails" => "4",
    "sql" => "14",
    # hidden
    "apache" => "22",
    "clojure" => "38",
    "d" => "26",
    "erlang" => "27",
    "fortran" => "28",
    "haskell" => "29",
    "ini" => "35",
    "io" => "24",
    "lisp" => "25",
    "lua" => "23",
    "makefile" => "31",
    "nu" => "36",
    "pascal" => "17",
    "puppet" => "39",
    "scala" => "32",
    "scheme" => "33",
    "smarty" => "34",
    "tex" => "37",
    # aliases
    "sh" => "13",
    "c" => "7",
    "c++" => "7",
    "objective-C" => "1",
    "objective-C++" => "1",
    "plain" => "6",
    "raw" => "6",
    "rails" => "4",
    "html" => "11",
    "xml" => "11",
    "js" => "10",
    "make" => "31",
);

sub uri { 'http://pastie.org/' }

sub fill_form {
    my $self = shift;
    my $mech = shift;
    my %args = @_;

    my $lang_id = $languages{"plain text"};
    $lang_id = $languages{lc($args{lang})}
        if (exists $args{lang} && exists $languages{lc($args{lang})});

    $mech->submit_form(
        fields => {
            "paste[body]"          => $args{text},
            "paste[authorization]" => 'burger', # set with JS to avoid bots
            "paste[restricted]"    => $args{private},
            "paste[parser_id]"     => $lang_id,
        },
    );
}

sub return {
    my $self = shift;
    my $mech = shift;

    my $prefix ='';
    my ($id) = $mech->title =~ /\#(\d+)/;
    if (!$id) {
        ($id) = $mech->content =~ m{http://pastie.org/\d+/wrap\?key=([a-z0-9]+)};
        $prefix = 'private/';
    }
    return (0, "Could not construct paste link.") if !$id;
    return (1, "http://pastie.org/$prefix$id");
}

1;
__END__

=pod

=head1 NAME

App::Nopaste::Service::Pastie - http://pastie.org

=cut
